source/common/upstream/cluster_manager_impl.cc

```diff
@@ -422,6 +422,17 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   // We have a situation that clusters will be immediately active, such as static and primary
   // cluster. So we must have this prevention logic here.
   if (cluster_data != warming_clusters_.end()) {
+    Network::TransportSocketFactory& factory =
+        cluster.info()->transportSocketMatcher().resolve(&cluster.info()->metadata()).factory_;
+    // If there is no secret entity, currently supports only TLS Certificate and Validation
+    // Context, when it failed to extract them via SDS, it will fail to change cluster status from
+    // warming to active. In current implementation, there is no strategy to activate clusters
+    // which failed to initialize at once.
+    // TODO(shikugawa): Consider retry strategy of clusters which failed to activate at once.
```

Comment:
This seems a bit dicey. If I'm reading the code/comments correctly, we have a race condition in cluster warming where a slow SDS update will leave a cluster stuck in the warming state forever.

Reply:
I can't imagine what problem will cause if warming state is being stuck forever. Sorry. It will be an expected behavior. What you say is that the retry comment is not required, isn't it?

Reply:
The issue is that this is very surprising behavior to an Envoy operator who experiences a race between SDS availability and cluster warming. In fact, I'd argue that this effectively breaks eventual warming semantics.

It looks like, from the comment, that the right long-term solution is to continue warming when SDS makes the necessary secret available.

I think we should have this behavior. If you don't want to do it in this PR, then we should at least put in place a runtime flag that disables this new behavior by default, and leave it disabled until we actually fix warming for folks who are relying on eventual consistency.

Reply:
Sure. So to speak, what you said is that we should implement to be available by keeping warming after no-available secret entity behavior occured, right? This feature is the blocker of next Istio release. So I overcome this by setting runtime feature flag temporary.